### PR TITLE
feat: introduce dockerChildPrefix option

### LINKED ISSUE #8238: Support unique naming schema for docker containers when running binarySource=docker
**What would you like Renovate to be able to do?**

Currently Renovate is naming docker containers based on the manager e.g. `renovate_python` for a python container.
This makes runnig multiple containers on a single host very hard without something like lxd or docker in docker o.s.

In a perfect world I would be able to run multiple renovate processes on the same node without them interfering with each other.

There is one aspect that needs to be considered.
What will happen to child containers when the process crashes?
In the current setup the next renovate run will detect the containers due to the global naming and cleans them up.
This aspect should also be supported with the new solution.

**Did you already have any implementation ideas?**

To find a good solution we need to clearify a few things:

* Renovate gives the container a certain name e.g. `renovate_python` for the `renovate/python` image.
* Renovate gives the container a certain label `renovate_child`.
* Renovate removes all containers with the label `renovate_child` at the start of any run.

With that in mind there are some ideas on how to implement this:

The simplest solution would be to add a global prefix config option that renovate would us to name the containers.
This would be easy to implement and can be set to e.g. the repository name.
This would require to also use the prefix in the `renovate_child` label so containers are not removed by any Renovate run by accident.
The drawback would be that a dangling container may never be removed when the same prefix is not used anymore.
To me this seems like acceptable drawback as long as this is also clearly communicated.

A more complex solution would include some kind of clean up routine.
E.g. remove all child container with a `renovate_child` label that are older than X days.

For the time being I would recommend (given the prefix is `foo_`):

* Add a `dockerChildPrefix` option that will add the `foo_` prefix to the label `renovate_child` and by default is empty, so the resulting label would be `foo_renovate_child`. 
* Add the prefix to the dangling cleanup function, so that the function will look only for images with the label `foo_renovate_child`.
* Add the prefix to the name of the docker image, so that the `renovate/python` image will be named `foo_renovate_docker`. This would prevent naming conflicts.
* Add documentation for the prefix and make clear that danglig containers are only removed on the next Renovate run with the same prefix.

If this is acceptable I would be happy to implement this feature in Renovate.

### PR DESCRIPTION
**Pull request by Chumper**
## Changes:

Introduces a new `dockerChildPrefix` option that influences the name and the label of docker sidecar containers.
This enables people to run Renovate as multiple processes on a single node without interfering with each others spawned docker containers.

## Context:

Closes #8238

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added unit tests, or
- [ ] Unit tests + ran on a real repository

### CODE REVIEW

**Comment by Chumper** [comment]
Needed to force push to fix my git config again...

File: docs/usage/self-hosted-configuration.md [hunk 1]
```diff
@@ -92,6 +92,12 @@ If you configure this to be different to the `baseDir`, it means you can have on
 
 Set to `false` to prevent usage of `--ignore-platform-reqs` in the Composer package manager.
 
+## dockerChildPrefix
+
+Override the default renovate sidecar Docker containers name prefix from `renovate_` to a custom value, so renovate will name and label images from a custom prefix.
+
+If this is set to `myprefix_` the final image name for `renovate/node` would be named `myprefix_renovate_node` instead of currently used `renovate_node` and be labeled `myprefix_child` instead of `renovate_child`.
```
Comment (rarkins) [inline docs/usage/self-hosted-configuration.md:99]: Is it really `renovate_node` by default or `renovate_renovate_node` ?

File: lib/config/common.ts [hunk 2]
```diff
@@ -80,6 +80,7 @@ export interface RenovateAdminConfig {
 
   dockerImagePrefix?: string;
   dockerUser?: string;
+  dockerChildPrefix?: string;
```
Comment (rarkins) [inline lib/config/common.ts:83]: Is it really optional if it has a default value?

File: lib/util/exec/index.ts [hunk 3]
```diff
@@ -34,7 +37,7 @@ export async function setExecConfig(
     execConfig[key] = value || null;
   }
   if (execConfig.binarySource === 'docker') {
-    await removeDanglingContainers();
+    await removeDanglingContainers(getContainerLabel(config.dockerChildPrefix));
```
Comment (rarkins) [inline lib/util/exec/index.ts:40]: Could this logic be pushed into the removeDanglingContainers() logic instead?

File: docs/usage/self-hosted-configuration.md (hunk 1 above)
Comment (Chumper) [reply docs/usage/self-hosted-configuration.md:99]: Previously we have not added any prefix to the container name so `renovate/node` would have been named `renovate_node` as container name and received `renovate_child` as label.

Now we prefix all container names with `renovate_` by default so `renovate/node` becomes `renovate_renovate_node` and label all container with `<prefix>child` so it becomes `renovate_child` as before.

Comment (Chumper) [reply docs/usage/self-hosted-configuration.md:99]: But yes, the documentation refers to the old style, I will adjust it.

File: lib/util/exec/index.ts (hunk 3 above)
Comment (Chumper) [reply lib/util/exec/index.ts:40]: We can do that, but we need access to the config somehow. If we don't pass it in, we need to make it global somehow.

File: lib/config/common.ts (hunk 2 above)
Comment (Chumper) [reply lib/config/common.ts:83]: That may be a misunderstanding on my part. 
If a default is set in the `definitions.ts` will the value always be present?
If so, then we can change it of course.

Comment (viceice) [reply lib/config/common.ts:83]:
```suggestion
  dockerChildPrefix: string;
```
🙃

File: lib/util/exec/common.ts [hunk 4]
```diff
@@ -14,6 +14,7 @@ export enum BinarySource {
 
 export interface ExecConfig {
   binarySource: Opt<BinarySource>;
+  dockerChildPrefix: Opt<string>;
```
Comment (viceice) [inline lib/util/exec/common.ts:17]:
```suggestion
  dockerChildPrefix: string;
```
Shouldn't be null or undefined.

File: lib/util/exec/docker/index.ts [hunk 5]
```diff
@@ -110,12 +110,23 @@ async function getDockerTag(
...
+    return prefix.concat(image.replace(/\//g, '_'));
+  }
+  return 'renovate_'.concat(image.replace(/\//g, '_'));
+}
+
+export function getContainerLabel(prefix: string): string {
+  if (prefix) {
+    return `${prefix}child`;
+  }
+  return 'renovate_child';
```
Comment (viceice) [inline lib/util/exec/docker/index.ts:124]:
```suggestion
export function getContainerName(image: string, prefix: string): string {
  return `${prefix}${image.replace(/\//g, '_'}`;
}

export function getContainerLabel(prefix: string): string {
    return `${prefix}child`;
```

Maybe we can inline this?

File: lib/util/exec/index.ts (hunk 3 above)
Comment (viceice) [reply lib/util/exec/index.ts:40]:
```suggestion
    await removeDanglingContainers(config.dockerChildPrefix);
```
I think @rarkins  prefer something like this. 🙃

File: lib/util/exec/docker/index.ts [hunk 6]
```diff
@@ -130,21 +141,20 @@ export async function removeDockerContainer(image: string): Promise<void> {
...
-      'Could not remove Docker container'
-    );
+    logger.info({ containerName, cmd }, 'Could not remove Docker container');
   }
 }
 
 // istanbul ignore next
-export async function removeDanglingContainers(): Promise<void> {
+export async function removeDanglingContainers(
+  childLabel: string
```
Comment (viceice) [inline lib/util/exec/docker/index.ts:154]:
```suggestion
  prefix: string
```

File: lib/util/exec/docker/index.ts [hunk 7]
```diff
@@ -130,21 +141,20 @@ export async function removeDockerContainer(image: string): Promise<void> {
...
+): Promise<void> {
   try {
-    const res = await rawExec(`docker ps --filter label=renovate_child -aq`, {
+    const res = await rawExec(`docker ps --filter label=${childLabel} -aq`, {
```
Comment (viceice) [inline lib/util/exec/docker/index.ts:157]:
```suggestion
    const res = await rawExec(`docker ps --filter label=${getContainerLabel(prefix)} -aq`, {
```

**Comment by Chumper** [comment]
I applied all suggestions.
This resulted in some snapshot to have `null` or `undefined` as prefix.
I fixed most of them except one.

@viceice  @rarkins Please review again

**Comment by Chumper** [comment]
I see that making the parameter required fails a lot of other locations.
I initially thought that it also is a prefix like the `dockerImagePrefix`.
In that case we could also hardcode the prefix as is done for the image prefix here: https://github.com/renovatebot/renovate/blob/master/lib/util/exec/docker/index.ts#L213

**Review by viceice — CHANGES_REQUESTED** [review]

**Comment by rarkins** [comment]
@Chumper sorry, this branch got conflicted after another PR merged. Can you update and resolve the conflicts?

**Comment by Chumper** [comment]
Rebased the branch

**Comment by Chumper** [comment]
@rarkins @viceice Do we need to discuss this further?

I think that after making the option optional we should be good to go. 
The only downside is that containers are now prefixed with `renovate_` which could lead to confusions.
I am happy to implement the option to be empty by default and make a case decision for the child labels.

when dockerChildPrefix = ``
container name: `renovate/node` => `renovate_node`
container label: `renovate_child`

when dockerChildPrefix = `foo`
container name: `renovate/node` => `foo_renovate_node`
container label: `renovate_foo_child`

when dockerChildPrefix = `renovate`
container name: `renovate/node` => `renovate_renovate_node`
container label: `renovate_renovate_child`

In that case the current behaviour is preserved and only when the prefix is set it differs.

**Comment by rarkins** [comment]
@Chumper I like your latest suggestion to default it to empty and retain backwards compatibility

**Comment by Chumper** [comment]
Superseded by #8613

**Pull request closed without merging**